rfd/0047-drop-vendor.md

```diff
@@ -0,0 +1,144 @@
+provide a versioned API module to external developers, while always building
+against latest for official builds. Unfortunately, replace directives don't play
+nice with vendoring, so we use an
+[error-prone symbolic link](https://github.com/gravitational/teleport/blob/30effc1f08b6a699772ff22f79ebe756fe1a1e34/Makefile#L942-L952)
+which is difficult to maintain in a cross-platform way, and has broken
+[gopls integration](https://github.com/gravitational/teleport/blob/30effc1f08b6a699772ff22f79ebe756fe1a1e34/Makefile#L942-L952)
+a common tool used in Go development environments.
+
+Lastly, there is no guarantee that the code committed to vendor actually
+reflects the contents of go.mod. The onus is on the developer to remember to run
+`make update-vendor` and commit the results after making changes to
+dependencies. This has created several cases of confusing build results amongst
+the dev team in recent history.
+
+In summary, dropping vendor would make go.mod/go.sum the single source of truth
+for our dependencies, reduce the size of the teleport repo, make code changes
+easier to review, and restore functionality with the latest version of `gopls`.
+
+### Security
+
+The removal of the vendor directory does eliminate one workflow that is nice
+from a security perspective and is worth discussing here. With vendoring, when a
+dependency is updated, the Teleport commit diff shows the actual source code
+changes to the dependency (or dependencies), allowing the developer to review
+the changes for bugs or even malicious code. If the vendor directory is removed,
+the developer would have to look at the release notes and code changes upstream
+when updating dependencies in order to determine whether or not to apply the update.
+
+While this "feature" sounds important in theory, in practice it is less useful.
+The diff has to be just the right size to attract developer attention but not so
+large as to disincentivize a review. In most cases, even with vendoring, a
+developer simply sees a large set of dependency changes and glances over them
+before moving on.
+
+Supply chain security is as important as ever these days, and while our current
+approach can (in some cases) make it easy for developers to audit changes, it
+still requires a disciplined development team and careful attention to detail
+and cannot be relied upon as the sole mechanism to prevent supply chain attacks.
+In fact, Go modules arguably offer better supply chain security, as modules are
+authenticated against an auditable checksum database. Some module proxies even
+offer vulnerability scanning of the modules they serve.
+
+Additionally, the Go project itself is making a number of improvements with
+respect to supply chain security. Starting in Go 1.18, the `go` command will
+embed additional version information in binaries, making it easier to generate a
+software bill-of-materials. The Go team is also working on a Go vulnerability
```

Comment:
Even with 1.17 (starting 1.15 AFAIR) you can call `go version -m <bin-path>` to get full list of dependencies that comes from go.mod. Funny enough this feature doesn't work for me for Apple M1 build. Apple x64 is ok 🤔